book/src/rfcs/ffadd.md

```diff
@@ -0,0 +1,192 @@
+
+For the second part of the gate, meaning the sum for the range check, the carry bits will remain unchanged. The reason behind this is the fact that $g > 0$ because $2^{264} > f$, and $r$ will be "positive" as well. That means, this sum follows the usual positive-addition carry structure. 
+
+## Gadget
+
+The foreign field gadget will be composed by 4 sets of `RangeCheck` gadgets for witnesses $a, b, r, s$ accounting for 16 rows in total; followed by one row with `ForeignFieldAdd` gate type; and a final `Zero` row. A total of 18 rows with 15 columns in Kimchi.
+
+| Row(s) | Gate type(s)        | Witness |
+| ------ | ------------------- | ------- |
+| 0-3    | `multi-range-check` | $a$     |
+| 4-7    | `multi-range-check` | $b$     |
+| 8-11   | `multi-range-check` | $r$     |
+| 12-15  | `multi-range-check` | $s$     |
+| 16     | `ForeignFieldAdd`   |         |
+| 17     | `Zero`              |         |
+
+
+### Layout
+
+For this gate, we need to perform 4 range checks to assert that the limbs of $a, b, r, s$ have a correct size, meaning they fit in $2^{88}$ (and thus, range-checking $a, b, r, s$ for $2^{264}$). Because each of these elements is split into 3 limbs, we will have to use 3 copy constraints between the `RangeCheck` gates and the `ForeignFieldAdd` rows (per element). That amounts to 12 copy constraints. Recalling that Kimchi only allows for the first 7 columns of each row to host a copy constraint, we necessarily have to use 2 rows for the actual addition gate. The layout of these two rows is the following:
+
+|            | Curr              | Next         |
+| ---------- | ----------------- | ------------ |
+| **Column** | `ForeignFieldAdd` | `Zero`       |
+| 0          | $a_0$ (copy)      | $r_0$ (copy) |
+| 1          | $a_1$ (copy)      | $r_1$ (copy) |
+| 2          | $a_2$ (copy)      | $r_2$ (copy) |
+| 3          | $b_0$ (copy)      | $u_0$ (copy) |
+| 4          | $b_1$ (copy)      | $u_1$ (copy) |
+| 5          | $b_2$ (copy)      | $u_2$ (copy) |
+| 6          | $x$               |
+| 7          | $c_0$             |
+| 8          | $c_1$             |
+| 9          | $k_0$             |
+| 10         | $k_1$             |
+| 11         |                   |
+| 12         |                   |
+| 13         |                   |
+| 14         |                   |
+
+### Constraints
+
+So far, we have pointed out the following sets of constraints:
+
+#### Main addition
+
+- $2^{88} \cdot c_0 = a_0 + b_0 - x \cdot f_0$
+- $2^{88} \cdot c_1 = a_1 + b_1 - x \cdot f_1 + c_0$
+- $2^{88} \cdot c_2 = a_2 + b_2 - x \cdot f_2 + c_1$
+
+#### Field check
+
+- $u_0 = r_0 + g_0 - k_0 \cdot 2^{88}$
+- $u_1 = r_1 + g_1 - k_1 \cdot 2^{88} + k_0$
+- $u_2 = r_2 + g_2 + k_1$
+
+#### Carry checks
+
+- $0 = c_0 \cdot (c_0 + 1) \cdot (c_0 - 1)$
+- $0 = c_1 \cdot (c_0 + 1) \cdot (c_1 - 1)$
+- $0 = c_2 \cdot (c_0 + 1) \cdot (c_2 - 1)$
+- $0 = k_0 \cdot (k_0 - 1)$ 
+- $0 = k_1 \cdot (k_1 - 1)$
```

Comment:
I think we need to check `q == 0` also

Reply:
oh no, not this one. This would mean no field overflow, but there can be.